### PR TITLE
Update AEAD limits for larger packets

### LINKED ISSUE #3701: The QUIC-TLS draft should define anti-forgery limits for packet lengths up to 2^16
This issue is in response to @chris-wood 's https://github.com/quicwg/base-drafts/pull/3620#issuecomment-633062212 .

PR #3620 introduces anti-forgery limits for packet lengths up to 2^14 as those were inherited from the maximum TLS record size. Since @chris-wood mentioned a way to compute limits for those sizes, we should use that and put the limit in the document, instead of leaving that as an exercise to the reader.

### PR DESCRIPTION
**Pull request by martinthomson**
The analysis describes conditions where higher limits might be allowed, and provides a formula in each case that can be used to determine other limits more readily.

The number of changes I've made here are somewhat frightening.  This is why I resisted doing this.  This will need careful double-checking.  I think that this is better, but I'm not sure that I feel happy about having spent the time.

The actual numbers don't concern me much, anyone can subtract 2, but the changes to the structure really need to be looked at.  Some of the changes are editorial, but they still need extra eyes.

Closes #3701.

### CODE REVIEW

**Comment by chris-wood** [comment]
cc @fxguenther

**Review by DavidSchinazi — APPROVED** [review]
This PR is exactly what I was hoping for, thanks for writing it. As far as I can tell, it's correct. It also fully addresses my concerns from #3701.

**Review by chris-wood — APPROVED** [review]
Thanks, @martinthomson! I checked the derivations and produced the same results.

**Review by MikeBishop — COMMENTED** [review]
The actual analysis is beyond my scope, but the text looks good.

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -2404,37 +2421,33 @@ no more than:
...
-
-~~~
-q <= 2^24.5
-~~~
-
-That is, endpoints cannot protect more than 2^23 packets with the same set of
-keys without causing an attacker to gain a larger advantage than the target of
-2^-57.  Note however that the integrity limits further constrain this value.
+The integrity limit in Theorem 1 in {{?CCM-ANALYSIS}} provides an attacker a
+strictly higher advantage for the same number of messages. As the targets
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-tls.md:2425]: I'm trying to say that the "value of the goal" is the same.  I think that this is correct.

I think that you might say "target advantages for A and B", but I think that "advantages" means something else and so I avoided the plural.

Maybe instead "As the target for both the confidentiality advantage and integrity advantage is the same, ..." ?

My grammar skillz aren't up to this, clearly.

[the remark this answers is not in the document]
Comment (MikeBishop) [reply draft-ietf-quic-tls.md:2425]: Or plural works in that last arrangement, so "As the targets for the confidentiality advantage and the integrity advantage are the same...."  I think any of them are fine, so long as they agree in number.  What stuck out to me was "targets advantage" more than anything else.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -2337,18 +2352,19 @@ distinguishing advantage between a real and random AEAD algorithm gained by an
 attacker is:
 
 ~~~
-2 * (q * l)^2 / 2^128
+2 * (q * l)^2 / 2^n
```
Comment (martinthomson) [inline draft-ietf-quic-tls.md:2355]: I forgot to call this out.  @chris-wood, can you check that this is right here?  I think that this comes from the `\sigma*B/2^n` term in Bad3 in [the paper](https://eprint.iacr.org/2018/993.pdf), but I want to make sure that my understanding is correct in that we are treating `\sigma*B` as `\sigma^2` in the single-user case.

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -2299,17 +2307,23 @@ o:
 : The amount of offline ideal cipher queries made by an adversary.
 
 The analyses that follow rely on a count of the number of block operations
-involved in producing each message. For simplicity, and to match the analysis of
-other AEAD functions in {{AEBounds}}, this analysis assumes a packet length of
-2^10 blocks; that is, a packet size limit of 2^14 bytes.
+involved in producing each message. This analysis is performed for packets of
+up to 2^11 (l = 2^7) and 2^16 (l = 2^12). A size of 2^11 is expected to be a
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2311]:
```suggestion
size up to 2^11 (l = 2^7) and 2^16 (l = 2^12). A size of 2^11 is expected to be a
```

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -2299,17 +2307,23 @@ o:
...
+limit that matches common deployment patterns, whereas the 2^16 is the maximum
+possible size of a QUIC packet. Only endpoints that strictly limit packet size
+can use the larger limits that are derived using the smaller packet size.
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2314]: limit is being used for multiple things here -- these limits and also packet size limit. 
```suggestion
can use the larger confidentiality and integrity limits that are derived using the smaller packet size.
```

File: draft-ietf-quic-tls.md [hunk 5]
```diff
@@ -2337,18 +2352,19 @@ distinguishing advantage between a real and random AEAD algorithm gained by an
...
+q <= 2^35 / l
 ~~~
 
-Thus, endpoints cannot protect more than 2^25 packets in a single connection
-without causing an attacker to gain an larger advantage than the target of
-2^-57.
+Thus, endpoints that do not send packets larger than 2^11 bytes cannot protect
+more than 2^28 packets in a single connection without causing an attacker to
+gain an larger advantage than the target of 2^-57. The limit for endpoints that
+do not limit packet size is instead 2^23.
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2367]:
```suggestion
allow for the packet size to be as large as 2^16 is instead 2^23.
```

File: draft-ietf-quic-tls.md [hunk 6]
```diff
@@ -2366,15 +2382,19 @@ term in this inequality dominates the rest, so the others can be removed without
...
 ~~~
-v <= 2^54
+v <= 2^64 / l
 ~~~
 
-For AEAD_AES_256_GCM, the second and fourth terms dominate the rest, so the
-others can be removed without affecting the result. This produces the following
-approximation:
+Endpoints that do not attempt to remove protection from packets larger than
+2^11 bytes can attempt to remove protection from at most 2^57 packets. Packets
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2389]:
```suggestion
2^11 bytes can attempt to remove protection from at most 2^57 packets. Endpoints
```

File: draft-ietf-quic-tls.md [hunk 7]
```diff
@@ -2366,15 +2382,19 @@ term in this inequality dominates the rest, so the others can be removed without
...
+attempt to remove protection from any packet can attempt to remove protection
+from at most 2^52 packets.
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2391]:
```suggestion
that do not restrict the size of processed packets can attempt to remove protection
from at most 2^52 packets.
```

File: draft-ietf-quic-tls.md [hunk 8]
```diff
@@ -2404,37 +2421,33 @@ no more than:
...
+v + q <= 2^34.5 / l
 ~~~
 
-Assuming `q = v`, endpoints cannot attempt to protect or authenticate more than
-2^23.5 packets with the same set of keys without causing an attacker to gain a
-larger advantage in forging packets than the target of 2^-57.
+By setting `q = v`, values for both confidentiality and integrity limits can be
+produced. Endpoints that limit packets to 2^11 bytes therefore have both
+confidentiality and integrity limits of 2^26.5 packets. Endpoints that do not
+limit packet size have a limit of 2^21.5.
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md:2450]:
```suggestion
restrict packet size have a limit of 2^21.5.
```

**Review by janaiyengar — APPROVED** [review]
A few minor points.

**Pull request merged**